subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java

```diff
@@ -124,7 +130,7 @@ public void apply(final Object target) {
                     return fileLookup.getFileResolver().getPatternSetFactory();
                 }
             };
-            services.add(ScriptPluginFactory.class, DefaultScriptPluginFactory.this);
+            services.add(ScriptPluginFactory.class, scriptPluginFactory);
```

Comment:
The service registry is specific for each `ScriptPluginImpl`. That's why we build it here. Each instance has a potentially different `ClassLoaderScope`.